benchmark/python/dot.py

```diff
@@ -83,198 +208,236 @@ def get_iter(path, data_shape, batch_size):
+    # Benchmark MXNet and Scipys dot operator
+    def bench_dot(lhs_shape, rhs_shape, lhs_stype, rhs_stype,
+                  lhs_den, rhs_den, trans_lhs, ctx, num_repeat=10, fw="mxnet", distribution="uniform"):
         set_default_context(ctx)
+        assert fw == "mxnet" or fw == "scipy"
+        # Set funcs
+        dot_func_sparse = mx.nd.dot if fw == "mxnet" else sp.spmatrix.dot
+        dot_func_dense = mx.nd.dot if fw == "mxnet" else np.dot
         # Create matrix instances
-        lhs_nd = rand_ndarray(lhs_shape, lhs_stype, density=lhs_den)
-        rhs_nd = rand_ndarray(rhs_shape, rhs_stype, density=rhs_den)
-        lhs_dns = lhs_nd if lhs_stype == 'default' else lhs_nd.tostype('default')
-        rhs_dns = rhs_nd if rhs_stype == 'default' else rhs_nd.tostype('default')
-        # One warm up run, verify correctness
-        out = mx.nd.dot(lhs_nd, rhs_dns, trans_lhs)
-        out_expected = mx.nd.dot(lhs_dns, rhs_dns, trans_lhs)
-        assert_almost_equal(out.asnumpy(), out_expected.asnumpy(), rtol=1e-2, atol=1e-3)
-        # Start benchmarking
+        lhs_nd = rand_ndarray(lhs_shape, lhs_stype, density=lhs_den, distribution=distribution)
+        # only uniform distribution supported for rhs
+        rhs_nd = rand_ndarray(rhs_shape, rhs_stype, density=rhs_den, distribution="uniform")
         lhs_nd.wait_to_read()
         rhs_nd.wait_to_read()
-        sparse_cost = measure_cost(repeat, mx.nd.dot, lhs_nd, rhs_nd, trans_lhs)
-        dense_cost = measure_cost(repeat, mx.nd.dot, lhs_dns, rhs_dns, trans_lhs)
-        speedup = dense_cost / sparse_cost
-        # Print results
-        m = lhs_shape[0]
-        k = lhs_shape[1]
-        n = rhs_shape[1]
-        results = '{:15.1f} {:15.1f} {:>10} {:8d} {:8d} {:8d} {:13.2f} {:13.2f} {:8.2f}'.format(lhs_den*100, rhs_den*100, str(ctx), m, k, n, sparse_cost*1000, dense_cost*1000, speedup)
-        print(results)
+        lhs_dns = None
+        rhs_dns = None
+        dense_cost = None
+        sparse_cost = None
+
+        if fw == "mxnet":
+            lhs_dns = lhs_nd if lhs_stype == 'default' else lhs_nd.tostype('default')
+            rhs_dns = rhs_nd if rhs_stype == 'default' else rhs_nd.tostype('default')
+            # One warm up run, verify correctness
+            out = dot_func_sparse(lhs_nd, rhs_dns, trans_lhs)
+            out_expected = dot_func_dense(lhs_dns, rhs_dns, trans_lhs)
+            assert_almost_equal(out.asnumpy(), out_expected.asnumpy(), rtol=1e-1, atol=1e-1)
+            sparse_cost = measure_cost(num_repeat, dot_func_sparse, lhs_nd, rhs_nd, trans_lhs)
+            dense_cost = measure_cost(num_repeat, dot_func_dense, lhs_dns, rhs_dns, trans_lhs)
+        else:
+            lhs_dns = np.transpose(lhs_nd.asnumpy()) if trans_lhs else lhs_nd.asnumpy()
+            rhs_dns = rhs_nd.asnumpy()
+            lhs_nd = sp.spmatrix.transpose(sp.csr_matrix(lhs_nd.asnumpy())) if trans_lhs else sp.csr_matrix(lhs_nd.asnumpy())
```

Comment:
sp.spmatrix.transpose just changes the view of the layout by reversing the dimensions. https://docs.scipy.org/doc/scipy-0.19.0/reference/generated/scipy.sparse.spmatrix.transpose.html
Yes I agree that in one case we are doing transpose early on but in mx.nd.dot we are doing transpose when we pass it to measure_cost. I can do the transpose before I call measure_cost for mxnet. Do you think this cost will be significant to affect benchmarking results ?